File: android/app/src/main/java/com/bayesmech/camalytics/coverage/CoverageTracker.kt
package com.bayesmech.camalytics.coverage

/**
 * Tracks data coverage over a rolling 10-second window.
 *
 * For each AR frame captured, records which data types were present.
 * Computes per-type coverage as a percentage of all frames in the window.
 * Thread-safe: called from IO dispatcher in ARDataCapture, read from UI thread.
 */
class CoverageTracker {

    private val windowMs = 10_000L

    private data class FrameRecord(
        val timestampMs: Long,
        val hasDepth: Boolean,
        val hasAccelerometer: Boolean,
        val hasGyroscope: Boolean,
        val hasMagnetometer: Boolean,
        val hasIntrinsics: Boolean,
        val hasPose: Boolean,
        val hasGeometry: Boolean,
        val hasGps: Boolean
    )

    private val lock = Any()
    private val frames = ArrayDeque<FrameRecord>()
    private var totalIntrinsicsCount = 0

    fun recordFrame(
        hasDepth: Boolean,
        hasAccelerometer: Boolean,
        hasGyroscope: Boolean,
        hasMagnetometer: Boolean,
        hasIntrinsics: Boolean,
        hasPose: Boolean,
        hasGeometry: Boolean,
        hasGps: Boolean
    ) {
        val now = System.currentTimeMillis()
        synchronized(lock) {
            if (hasIntrinsics) totalIntrinsicsCount++
            frames.addLast(
                FrameRecord(
                    now, hasDepth, hasAccelerometer, hasGyroscope,
                    hasMagnetometer, hasIntrinsics, hasPose, hasGeometry, hasGps
                )
            )
            pruneOldFrames(now)
        }
    }

    fun getStats(): CoverageStats {
        val now = System.currentTimeMillis()
        synchronized(lock) {
            pruneOldFrames(now)
            val total = frames.size
            if (total == 0) return CoverageStats(cameraIntrinsicsCount = totalIntrinsicsCount)

            val timeSpanMs = if (total >= 2) {
                frames.last().timestampMs - frames.first().timestampMs
            } else {
                1000L
            }
            val fps = if (timeSpanMs > 0) (total * 1000f / timeSpanMs) else 0f

            return CoverageStats(
                depthCoverage = frames.count { it.hasDepth } * 100f / total,
                accelerometerCoverage = frames.count { it.hasAccelerometer } * 100f / total,
                gyroscopeCoverage = frames.count { it.hasGyroscope } * 100f / total,
                magnetometerCoverage = frames.count { it.hasMagnetometer } * 100f / total,
                cameraIntrinsicsCount = totalIntrinsicsCount,
                poseCoverage = frames.count { it.hasPose } * 100f / total,
                inferredGeometryCoverage = frames.count { it.hasGeometry } * 100f / total,
                gpsCoverage = frames.count { it.hasGps } * 100f / total,
                averageFps = fps
            )
        }
    }

    fun reset() {
        synchronized(lock) {
            frames.clear()
            totalIntrinsicsCount = 0
        }
    }

    private fun pruneOldFrames(now: Long) {
        while (frames.isNotEmpty() && now - frames.first().timestampMs > windowMs) {
            frames.removeFirst()
        }
    }
}

File: ios/BayesMechVision/Coverage/CoverageTracker.swift
import Foundation

final class CoverageTracker: @unchecked Sendable {
    private struct FrameRecord {
        let timestamp: Date
        let hasDepth: Bool
        let hasAccelerometer: Bool
        let hasGyroscope: Bool
        let hasMagnetometer: Bool
        let hasIntrinsics: Bool
        let hasPose: Bool
        let hasGeometry: Bool
        let hasGPS: Bool
    }

    private let lock = NSLock()
    private let window: TimeInterval = 10
    private var frames: [FrameRecord] = []
    private var totalIntrinsicsCount = 0

    func recordFrame(
        hasDepth: Bool,
        hasAccelerometer: Bool,
        hasGyroscope: Bool,
        hasMagnetometer: Bool,
        hasIntrinsics: Bool,
        hasPose: Bool,
        hasGeometry: Bool,
        hasGPS: Bool
    ) {
        lock.lock()
        defer { lock.unlock() }

        if hasIntrinsics {
            totalIntrinsicsCount += 1
        }

        let record = FrameRecord(
            timestamp: Date(),
            hasDepth: hasDepth,
            hasAccelerometer: hasAccelerometer,
            hasGyroscope: hasGyroscope,
            hasMagnetometer: hasMagnetometer,
            hasIntrinsics: hasIntrinsics,
            hasPose: hasPose,
            hasGeometry: hasGeometry,
            hasGPS: hasGPS
        )

        frames.append(record)
        pruneOldFrames()
    }

    func stats() -> CoverageStats {
        lock.lock()
        defer { lock.unlock() }

        pruneOldFrames()
        guard !frames.isEmpty else {
            return CoverageStats(cameraIntrinsicsCount: totalIntrinsicsCount)
        }

        let total = Float(frames.count)
        let span: TimeInterval
        if let first = frames.first?.timestamp, let last = frames.last?.timestamp {
            span = max(last.timeIntervalSince(first), 1.0)
        } else {
            span = 1.0
        }

        return CoverageStats(
            depthCoverage: percentage(\.hasDepth, total: total),
            accelerometerCoverage: percentage(\.hasAccelerometer, total: total),
            gyroscopeCoverage: percentage(\.hasGyroscope, total: total),
            magnetometerCoverage: percentage(\.hasMagnetometer, total: total),
            cameraIntrinsicsCount: totalIntrinsicsCount,
            poseCoverage: percentage(\.hasPose, total: total),
            inferredGeometryCoverage: percentage(\.hasGeometry, total: total),
            gpsCoverage: percentage(\.hasGPS, total: total),
            averageFPS: Float(Double(frames.count) / span)
        )
    }

    func reset() {
        lock.lock()
        defer { lock.unlock() }
        frames.removeAll()
        totalIntrinsicsCount = 0
    }

    private func percentage(_ keyPath: KeyPath<FrameRecord, Bool>, total: Float) -> Float {
        guard total > 0 else { return 0 }
        let count = frames.filter { $0[keyPath: keyPath] }.count
        return Float(count) * 100.0 / total
    }

    private func pruneOldFrames() {
        let cutoff = Date().addingTimeInterval(-window)
        frames.removeAll { $0.timestamp < cutoff }
    }
}
